# Assignment 5: Github Project
Name: Claire Anne Gaspar

### Olympic 2024 Data Merge

This script merges two datasets: 

* "Olympic2024_Medallists.csv" - A dataset containing the medal winners.
* "Olympic2024_Athletes.csv" - A dataset with general athlete information.

Objective:
By merging these datasets, we can enrich the medalist data with additional athlete details such as age, country, and sport.

Merge Key:
Both datasets share a common key: 'Athlete Code', which will be used to combine the information.

Output:
The merged dataset will be saved as "Olympics2024_MergedData.csv" in the clean_data folder.

### 1. Import Libraries

Before merging the datasets, we need to import pandas.

In [4]:
import pandas as pd

### 2. Load Dataset

First, load the first dataset and check its structure. 

In [6]:
# Load the first dataset (Medallists)

medallists_data = pd.read_csv("../data/Olympic2024_Medallists.csv")

In [7]:
medallists_data.shape

medallists_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2315 entries, 0 to 2314
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   medal_date    2315 non-null   object
 1   medal_type    2315 non-null   object
 2   medal_code    2315 non-null   int64 
 3   name          2312 non-null   object
 4   gender        2312 non-null   object
 5   country_code  2312 non-null   object
 6   country       2312 non-null   object
 7   country_long  2312 non-null   object
 8   nationality   2310 non-null   object
 9   team          1555 non-null   object
 10  team_gender   1555 non-null   object
 11  discipline    2315 non-null   object
 12  event         2315 non-null   object
 13  event_type    2315 non-null   object
 14  url_event     2294 non-null   object
 15  birth_date    2312 non-null   object
 16  code_athlete  2315 non-null   int64 
 17  code_team     1555 non-null   object
dtypes: int64(2), object(16)
memory usage: 325.7+ KB


In [8]:
medallists_data.head(3)

medal_date    medal_type  medal_code             name gender country_code  \
0  2024-07-27    Gold Medal           1  EVENEPOEL Remco   Male          BEL   
1  2024-07-27  Silver Medal           2    GANNA Filippo   Male          ITA   
2  2024-07-27  Bronze Medal           3    van AERT Wout   Male          BEL   

   country country_long nationality team team_gender    discipline  \
0  Belgium      Belgium     Belgium  NaN         NaN  Cycling Road   
1    Italy        Italy       Italy  NaN         NaN  Cycling Road   
2  Belgium      Belgium     Belgium  NaN         NaN  Cycling Road   

                         event event_type  \
0  Men's Individual Time Trial        ATH   
1  Men's Individual Time Trial        ATH   
2  Men's Individual Time Trial        ATH   

                                           url_event  birth_date  \
0  /en/paris-2024/results/cycling-road/men-s-indi...  2000-01-25   
1  /en/paris-2024/results/cycling-road/men-s-indi...  1996-07-25   
2  /en/paris-2024/results/cycling-road/men-s-indi...  1994-09-15   

   code_athlete code_team  
0       1903136       NaN  
1       1923520       NaN  
2       1903147       NaN

Then, load the second dataset and check its structure as well. 

In [10]:
# Load the second dataset (Athletes)

athletes_data = pd.read_csv("../data/Olympic2024_Athletes.csv")

In [11]:
athletes_data.shape

athletes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11110 entries, 0 to 11109
Data columns (total 35 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   code                11110 non-null  int64  
 1   name                11110 non-null  object 
 2   name_short          11110 non-null  object 
 3   name_tv             11110 non-null  object 
 4   gender              11110 non-null  object 
 5   function            11110 non-null  object 
 6   country_code        11110 non-null  object 
 7   country             11110 non-null  object 
 8   country_long        11110 non-null  object 
 9   nationality         11107 non-null  object 
 10  nationality_full    11107 non-null  object 
 11  nationality_code    11107 non-null  object 
 12  height              11110 non-null  int64  
 13  weight              11108 non-null  float64
 14  disciplines         11110 non-null  object 
 15  events              11110 non-null  object 
 16  birt

In [12]:
athletes_data.head(3)

code              name    name_short           name_tv gender function  \
0  1532872  ALEKSANYAN Artur  ALEKSANYAN A  Artur ALEKSANYAN   Male  Athlete   
1  1532873    AMOYAN Malkhas      AMOYAN M    Malkhas AMOYAN   Male  Athlete   
2  1532874   GALSTYAN Slavik    GALSTYAN S   Slavik GALSTYAN   Male  Athlete   

  country_code  country country_long nationality  ...  \
0          ARM  Armenia      Armenia     Armenia  ...   
1          ARM  Armenia      Armenia     Armenia  ...   
2          ARM  Armenia      Armenia     Armenia  ...   

                      family                        lang  \
0  Father, Gevorg Aleksanyan  Armenian, English, Russian   
1                        NaN                    Armenian   
2                        NaN                    Armenian   

                                               coach  \
0                    Gevorg Aleksanyan (ARM), father   
1                                                NaN   
2  Personal: Martin Alekhanyan (ARM).<br>National...   

                                              reason  \
0  He followed his father and his uncle into the ...   
1                                                NaN   
2                                                NaN   

                                                hero  \
0  Footballer Zinedine Zidane (FRA), World Cup wi...   
1                                                NaN   
2                                                NaN   

                       influence  \
0  His father, Gevorg Aleksanyan   
1                            NaN   
2                            NaN   

                                          philosophy  \
0  "Wrestling is my life." (mediamax.am. 18 May 2...   
1  "To become a good athlete, you first have to b...   
2                                                NaN   

                                  sporting_relatives ritual other_sports  
0                                                NaN    NaN          NaN  
1  Uncle, Roman Amoyan (wrestling), 2008 Olympic ...    NaN          NaN  
2                                                NaN    NaN          NaN  

[3 rows x 35 columns]

### 3. Check and Clean the Data

Rename code to code_athlete in athletes_data to allow merging.

In [14]:
## Rename the column in athletes_data to match medallists_data

athletes_data.rename(columns={"code": "code_athlete"}, inplace=True)

### 4. Merge the Data

In [16]:
##  Now that both datasets are cleaned and have a common key, we are ready to merge them.

merged_datasets = pd.merge(medallists_data, athletes_data, on="code_athlete", how="left")

In [17]:
merged_datasets.head(3)

medal_date    medal_type  medal_code           name_x gender_x  \
0  2024-07-27    Gold Medal           1  EVENEPOEL Remco     Male   
1  2024-07-27  Silver Medal           2    GANNA Filippo     Male   
2  2024-07-27  Bronze Medal           3    van AERT Wout     Male   

  country_code_x country_x country_long_x nationality_x team  ...  \
0            BEL   Belgium        Belgium       Belgium  NaN  ...   
1            ITA     Italy          Italy         Italy  NaN  ...   
2            BEL   Belgium        Belgium       Belgium  NaN  ...   

                                              family                    lang  \
0       Wife, Oumaima. Father, Patrick. Mother, Agna          Dutch, English   
1  Fiancee, Carlotta Morino. Father, Marco. Mothe...        English, Italian   
2  Wife, Sarah (married 2018). Son, Georges (2021...  Dutch, English, French   

                                               coach  \
0                                                NaN   
1  Personal: Dario Cioni (ITA). <br>Personal (tim...   
2                                                NaN   

                                              reason  \
0  He initially played football but decided to fo...   
1  His grandparents gave him his first bike and f...   
2  His father was an amateur racer and he grew up...   

                                                hero    influence  \
0  Cyclist Philippe Gilbert (BEL), three-time Oly...  His parents   
1  Cyclist Fabian Cancellara (SUI), two-time Olym...          NaN   
2                                                NaN          NaN   

                                          philosophy  \
0                                                NaN   
1                                                NaN   
2  "The biggest thing is that I never limited mys...   

                                  sporting_relatives ritual  \
0  Father, Patrick (road cycling), winner of the ...    NaN   
1  Father, Marco Ganna (canoe sprint), competed f...    NaN   
2                                                NaN    NaN   

                                        other_sports  
0  He played football at youth international leve...  
1                                                NaN  
2  Cyclo-cross: Won gold at the world championshi...  

[3 rows x 52 columns]

In [18]:
merged_datasets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2315 entries, 0 to 2314
Data columns (total 52 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   medal_date          2315 non-null   object 
 1   medal_type          2315 non-null   object 
 2   medal_code          2315 non-null   int64  
 3   name_x              2312 non-null   object 
 4   gender_x            2312 non-null   object 
 5   country_code_x      2312 non-null   object 
 6   country_x           2312 non-null   object 
 7   country_long_x      2312 non-null   object 
 8   nationality_x       2310 non-null   object 
 9   team                1555 non-null   object 
 10  team_gender         1555 non-null   object 
 11  discipline          2315 non-null   object 
 12  event               2315 non-null   object 
 13  event_type          2315 non-null   object 
 14  url_event           2294 non-null   object 
 15  birth_date_x        2312 non-null   object 
 16  code_a

## Why Choose how="left"?

We are using a left join because our primary focus is on the medal winners.
This ensures that all medalists remain in the dataset, even if some details from the athletes_data are missing.

Key reasons for using Left Join
* The medallists_data contains all Olympic 2024 medalists.
* Using a left join ensures that all medalists stay, even if their details are missing from athletes_data.
* If an athlete exists in medallists_data but not in athletes_data, their personal details (e.g., height, weight, birthdate) will be NaN (missing values) instead of being dropped.
  
### What if we used an Inner Join (how="inner")?

Problem: Medalists without matching athlete records would be removed from the final dataset.
Inner Join is only useful if we only want athletes who won medals and have full details.

Since our main focus is to keep all medalists, an inner join is too restrictive and would result in missing some medalists from our dataset.

### Why not Right or Outer?
  
* Right Join (how="right"):
    * Would keep all athletes, even those who didn’t win medals.
    * This would make the dataset too large and include unnecessary data for medal analysis.

* Outer Join (how="outer"):
    * Would keep all medalists and all athletes, even if they don’t match.
    * This would introduce many NaN (missing) values, making the dataset harder to work with.

In [20]:
print(merged_datasets.isnull().sum())


medal_date               0
medal_type               0
medal_code               0
name_x                   3
gender_x                 3
country_code_x           3
country_x                3
country_long_x           3
nationality_x            5
team                   760
team_gender            760
discipline               0
event                    0
event_type               0
url_event               21
birth_date_x             3
code_athlete             0
code_team              760
name_y                   3
name_short               3
name_tv                  3
gender_y                 3
function                 3
country_code_y           3
country_y                3
country_long_y           3
nationality_y            4
nationality_full         4
nationality_code         4
height                   3
weight                   4
disciplines              3
events                   3
birth_date_y             3
birth_place            329
birth_country          230
residence_place        838
r

### Analysis of Missing Values in Merged Dataset

After merging two datasets using a left join (how="left"), we checked for missing (NaN) values in the merged dataset. Below is an analysis of where these missing values come from and why they appear.

#### 1. Missing Values That Already Existed in medallists_data
* Some missing values in name_x, gender_x, and birth_date_x were already present before merging.
* This means these medalists were missing information even in the original dataset, so the merge did not create these `NaN` values.

#### 2. Missing Values Created by the Merge (NaN from athletes_data)
Since we used a left join, all medalists remain in the dataset, but:

* If a medalist had no match in athletes_data, their details appear as NaN (e.g., height, weight, birth_place, birth_country).
* Fields from athletes_data have higher NaN counts because not all medalists were found in the athletes dataset.
  - Example: birth_place has 329 missing values.
  - Example: hobbies, philosophy, rituals have 1000+ missing values, likely because they were optional fields.



### 5. Save Merged Dataset

In [25]:
merged_datasets.to_csv("../clean_data/Olympics2024_MergedData.csv", index=False)